### PR TITLE
FW-526. SCuM port.

### PR DESCRIPTION
**Pull request by changtengfei**
Implemented project list:
- 01bsp_bsp_timer
- 01bsp_uart
- 01bsp_leds
- 01bsp_radiotimer
- 01bsp_radio
- 02drv_openserial
- 02drv_opentimers
- 03oos_macpong
- 03oos_openwsn

### CODE REVIEW

File: bsp/boards/scum/board.c [hunk 1]
```diff
@@ -0,0 +1,53 @@
+/**
+\brief GINA-specific definition of the "board" bsp module.
```
Comment (twatteynelinear) [inline bsp/boards/scum/board.c]: GINA -> SCuM

Comment (twatteyne) [inline bsp/boards/scum/board.c]: GINA -> SCuM

File: bsp/boards/scum/cm0dsasm.s [hunk 2]
```diff
@@ -0,0 +1,153 @@
+Stack_Size      EQU     0x0800                            ; 4KB of STACK
+
```
Comment (twatteyne) [inline bsp/boards/scum/cm0dsasm.s]: are all these newlines needed? Could you please remove them?

File: bsp/boards/scum/debugpins.c [hunk 3]
```diff
@@ -0,0 +1,93 @@
+/**
+\brief SCuM-specific definition of the "debugpins" bsp module.
+
+To be implemented after issue: SCUM-25
```
Comment (twatteyne) [inline bsp/boards/scum/debugpins.c]: remove from here

File: bsp/boards/scum/debugpins.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+//=========================== defines =========================================
+
+//=========================== variables =======================================
+
+//=========================== prototypes ======================================
+
+//=========================== public ==========================================
+
+void debugpins_init() {
+}
```
Comment (twatteyne) [inline bsp/boards/scum/debugpins.c]: in each function add the following comment:

```
empty for now, see https://openwsn.atlassian.net/browse/SCUM-25
```

File: bsp/boards/scum/eui64.c [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+*/
+
+#include "string.h"
+#include "eui64.h"
+
+//=========================== defines =========================================
+
+//=========================== variables =======================================
+
+const uint8_t eui64[8] = {0x00,0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77};
```
Comment (twatteyne) [inline bsp/boards/scum/eui64.c]: please create an issue about the need for an EUI-64 address on SCuM

File: bsp/boards/scum/leds.c [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    GPIO_REG__OUTPUT    &= ~0x04;
+}
+void    leds_sync_toggle() {
+    GPIO_REG__OUTPUT    ^=  0x04;
+}
+uint8_t leds_sync_isOn() {
+    return (uint8_t)(GPIO_REG__OUTPUT & 0x04)>>2;
+}
+
+// yellow
```
Comment (twatteyne) [inline bsp/boards/scum/leds.c]: these color labels seem wrong. please update with name of the LED on the silkscreen.

File: bsp/boards/scum/radio.c [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+#include "board.h"
+#include "radio.h"
+#include "radiotimer.h"
+#include "debugpins.h"
+#include "leds.h"
+#include "memory_map.h"
+
+//=========================== defines =========================================
+
+#define LENGTH_MAX_RX_BUFFER  130 //// 1B spi address, 1B length, 125B data, 2B CRC, 1B LQI
```
Comment (twatteyne) [inline bsp/boards/scum/radio.c]: comment seems wrong, there is no SPI.

File: bsp/boards/scum/radio.c [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+void radio_init() {
+
+    // clear variables
+    memset(&radio_vars,0,sizeof(radio_vars_t));
+    
+    // change state
+    radio_vars.state                = RADIOSTATE_STOPPED;
+    
+    // Enable all interrupts and pulses to radio timer: no loaddone interrupt
+    RFCONTROLLER_REG__INT_CONFIG    = 0x3FE;
```
Comment (twatteyne) [inline bsp/boards/scum/radio.c]: replace magic values with `#defines` `|`'ed together

File: bsp/boards/scum/radio.c [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+void radio_getReceivedFrame(uint8_t* pBufRead,
+                            uint8_t* pLenRead,
+                            uint8_t  maxBufLen,
+                             int8_t* pRssi,
+                            uint8_t* pLqi,
+                               bool* pCrc) {
+    
+    //===== crc
+    *pCrc           = 1;
```
Comment (twatteyne) [inline bsp/boards/scum/radio.c]: magic value. comment and `#define`

File: bsp/boards/scum/radio.c [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+   
+    //===== rssi
+    *pRssi          = -91;
```
Comment (twatteyne) [inline bsp/boards/scum/radio.c]: magic value. comment and `#define`

File: bsp/boards/scum/memory_map.h [hunk 11]
```diff
@@ -0,0 +1,72 @@
+//AHB Peripheral
```
Comment (twatteyne) [inline bsp/boards/scum/memory_map.h]: add header with Sahar and you as authors

File: bsp/boards/scum/memory_map.h [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+#define RFTIMER_REG__COMPARE5_CONTROL   *(unsigned int*)(AHB_RFTIMER_BASE + 0x44)
+#define RFTIMER_REG__COMPARE6_CONTROL   *(unsigned int*)(AHB_RFTIMER_BASE + 0x48)
+#define RFTIMER_REG__COMPARE7_CONTROL   *(unsigned int*)(AHB_RFTIMER_BASE + 0x4C)
+#define RFTIMER_REG__CAPTURE0           *(unsigned int*)(AHB_RFTIMER_BASE + 0x50)
+#define RFTIMER_REG__CAPTURE1           *(unsigned int*)(AHB_RFTIMER_BASE + 0x54)
+#define RFTIMER_REG__CAPTURE2           *(unsigned int*)(AHB_RFTIMER_BASE + 0x58)
+#define RFTIMER_REG__CAPTURE3           *(unsigned int*)(AHB_RFTIMER_BASE + 0x5C)
+#define RFTIMER_REG__CAPTURE0_CONTROL   *(unsigned int*)(AHB_RFTIMER_BASE + 0x60)
+#define RFTIMER_REG__CAPTURE1_CONTROL   *(unsigned int*)(AHB_RFTIMER_BASE + 0x64)
+#define RFTIMER_REG__CAPTURE2_CONTROL   *(unsigned int*)(AHB_RFTIMER_BASE + 0x68)
```
Comment (twatteyne) [inline bsp/boards/scum/memory_map.h]: many more `#define`s needed for each register bit

File: bsp/boards/scum/radio.c [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+kick_scheduler_t radio_isr() {
+    
+    PORT_TIMER_WIDTH capturedTime;
+    PORT_TIMER_WIDTH irq_status = RFCONTROLLER_REG__INT;
+    PORT_TIMER_WIDTH irq_error  = RFCONTROLLER_REG__ERROR;
+    
+    capturedTime                = radiotimer_getCapturedTime();
+    
+    if (irq_status & 0x00000002 || irq_status & 0x00000008){
+        // change state
```
Comment (twatteyne) [inline bsp/boards/scum/radio.c]: add comment about what this means

File: bsp/boards/scum/radio.c [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+            radio_vars.startFrame_cb(capturedTime);
+            // clear interruption bit
+            RFCONTROLLER_REG__INT_CLEAR = irq_status;
+            // kick the OS
+            return KICK_SCHEDULER;
+        }
+    }
+    
+    if (irq_status & 0x00000004 || irq_status & 0x00000010){
+        // change state
```
Comment (twatteyne) [inline bsp/boards/scum/radio.c]: add comment about what this means

File: bsp/boards/scum/radio.c [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+            // kick the OS
+            return KICK_SCHEDULER;
+        } else {
+            while(1);
+        }
+    }
+    
+    if (irq_error == 0) {
+        
+        // print out the error here.
```
Comment (twatteyne) [inline bsp/boards/scum/radio.c]: is this a TODO?

File: bsp/boards/scum/radiotimer.c [hunk 16]
```diff
@@ -0,0 +1,135 @@
+/**
+\brief SCuM-specific definition of the "radiotimer" bsp module.
+
+\author Tengfei Chang <tengfei.chang@inria.fr>, Auguest 2016.
```
Comment (twatteyne) [inline bsp/boards/scum/radiotimer.c]: Auguest -> August

File: bsp/boards/scum/radiotimer.c [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    while(1);
+}
+
+void radiotimer_setEndFrameCb(radiotimer_capture_cbt cb) {
+    while(1);
+}
+
+void radiotimer_start(PORT_RADIOTIMER_WIDTH period) {
+    // set period of radiotimer
+    RFTIMER_REG__MAX_COUNT          = period*61/4;
```
Comment (twatteyne) [inline bsp/boards/scum/radiotimer.c]: why the 61? please add comment
I would be very careful about using different operations on a single line. create a temp variable on which you do the operations, then apply to register

File: bsp/boards/scum/radiotimer.c [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    // set compare timer counter 0 to perform an overflow interrupt
+    RFTIMER_REG__COMPARE1           = 0;
+    // enable compare0 module and interrup
+    RFTIMER_REG__COMPARE1_CONTROL   = 0x03;
+}
+
+//===== direct access
+
+PORT_RADIOTIMER_WIDTH radiotimer_getValue() {
+    return RFTIMER_REG__COUNTER*4/61;
```
Comment (twatteyne) [inline bsp/boards/scum/radiotimer.c]: same as above.

File: bsp/boards/scum/uart.c [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+void    uart_writeByte(uint8_t byteToWrite){
+    UART_REG__TX_DATA = byteToWrite;
+    // there is no txdone interruption, call the handler directly
+    uart_tx_isr();
+}
+
+void    uart_enableInterrupts(){ 
+
```
Comment (twatteyne) [inline bsp/boards/scum/uart.c]: UART interrupts can be enabled/disabled, no?

Comment (changtengfei) [inline bsp/boards/scum/uart.c]: no, there is not register to enable or disable UART interrupt. It's enabled directly.

File: bsp/boards/scum/board.c (hunk 1 above)
Comment (changtengfei) [inline bsp/boards/scum/board.c]: Done.

File: bsp/boards/scum/cm0dsasm.s (hunk 2 above)
Comment (changtengfei) [inline bsp/boards/scum/cm0dsasm.s]: removed.

File: bsp/boards/scum/debugpins.c (hunk 3 above)
Comment (changtengfei) [inline bsp/boards/scum/debugpins.c]: removed.

File: bsp/boards/scum/debugpins.c (hunk 4 above)
Comment (changtengfei) [inline bsp/boards/scum/debugpins.c]: added.

File: bsp/boards/scum/eui64.c (hunk 5 above)
Comment (changtengfei) [inline bsp/boards/scum/eui64.c]: created at: https://openwsn.atlassian.net/browse/SCUM-29

File: bsp/boards/scum/leds.c (hunk 6 above)
Comment (changtengfei) [inline bsp/boards/scum/leds.c]: updated.

File: bsp/boards/scum/memory_map.h (hunk 11 above)
Comment (changtengfei) [inline bsp/boards/scum/memory_map.h]: Done.

File: bsp/boards/scum/memory_map.h (hunk 12 above)
Comment (changtengfei) [inline bsp/boards/scum/memory_map.h]: They are defined in each modules.  e.g. the bit related to radiotimer is defined in radiotimer.c file. The same with radio.c file.

File: bsp/boards/scum/radio.c (hunk 7 above)
Comment (changtengfei) [inline bsp/boards/scum/radio.c]: Fixed.

File: bsp/boards/scum/radio.c (hunk 8 above)
Comment (changtengfei) [inline bsp/boards/scum/radio.c]: replaced by register bits combination.

File: bsp/boards/scum/radio.c (hunk 9 above)
Comment (changtengfei) [inline bsp/boards/scum/radio.c]: defined.

File: bsp/boards/scum/radio.c (hunk 10 above)
Comment (changtengfei) [inline bsp/boards/scum/radio.c]: defined.

File: bsp/boards/scum/radio.c (hunk 13 above)
Comment (changtengfei) [inline bsp/boards/scum/radio.c]: added.

File: bsp/boards/scum/radio.c (hunk 14 above)
Comment (changtengfei) [inline bsp/boards/scum/radio.c]: added.

File: bsp/boards/scum/radio.c (hunk 15 above)
Comment (changtengfei) [inline bsp/boards/scum/radio.c]: yes, added more comments and TBD there.

File: bsp/boards/scum/radiotimer.c (hunk 16 above)
Comment (changtengfei) [inline bsp/boards/scum/radiotimer.c]: Fixed, thanks!

File: bsp/boards/scum/radiotimer.c (hunk 17 above)
Comment (changtengfei) [inline bsp/boards/scum/radiotimer.c]: This is used to convert the timer counter from 500KHz to 32768 which is used in the protocol. They are replaced by marco defines and more more commands are added.

File: bsp/boards/scum/radiotimer.c (hunk 18 above)
Comment (changtengfei) [inline bsp/boards/scum/radiotimer.c]: Fixed.

File: bsp/boards/scum/board_info.h [hunk 20]
```diff
@@ -30,6 +30,23 @@ to return the board's description.
 #define SCHEDULER_WAKEUP()                  
 #define SCHEDULER_ENABLE_INTERRUPT()        
 
+// ==== SCuM RF timer specific 
+
+// since SCuM uses 500KHz, but the statck protocol is designed for 32KHz.
```
Comment (twatteyne) [inline bsp/boards/scum/board_info.h]: please proof read this comment and fix the typos

File: bsp/boards/scum/board_info.h [hunk 21]
```diff
@@ -30,6 +30,23 @@ to return the board's description.
 #define SCHEDULER_WAKEUP()                  
 #define SCHEDULER_ENABLE_INTERRUPT()        
 
+// ==== SCuM RF timer specific 
+
+// since SCuM uses 500KHz, but the statck protocol is designed for 32KHz.
+// the following two marco are used to convert the couter value bewteen
+// two different frequency clocks. 500000/32768 is approximately 61/4.
+#define TIMER_COUTER_CONVERT_32K_TO_500K(value)    value*61/4
```
Comment (twatteyne) [inline bsp/boards/scum/board_info.h]: I'm a bit afraid of these several operations in one line. We can try, but could you please add parenthesis so it's clear in what order you want the division and multiplication to happen?

File: bsp/boards/scum/memory_map.h [hunk 22]
```diff
@@ -0,0 +1,79 @@
+/**
+\brief SCuM board's addresses of the memory-mapped registers .
+
+\author Sahar Misha     <smesri@berkeley.edu>,      August 2016.
```
Comment (twatteyne) [inline bsp/boards/scum/memory_map.h]: please fix typo in Sahar's last name

File: bsp/boards/scum/memory_map.h (hunk 12 above)
Comment (twatteyne) [inline bsp/boards/scum/memory_map.h]: Hmm. I'm a little bit puzzled by this. In an ideal world, the HW team would generate this one file automatically, running a little tool on their HW design. So we want all the register-related `#define`s to be in  the same file. Is it possible to copy-paste things back to this file?

File: bsp/boards/scum/radio.c [hunk 23]
```diff
@@ -0,0 +1,312 @@
...
+#define TX_LOAD_DONE_INT        0x001
+#define TX_SFD_DONE_INT         0x002
+#define TX_SEND_DONE_INT        0x004
+#define RX_SFD_DONE_INT         0x008
+#define RX_DONE_INT             0x010
+
+// ==== default crc check result and rssi value
+
+#define DEFAULT_CRC_CHECK       0x01
+#define DEFAULT_RSSI            -91
```
Comment (twatteyne) [inline bsp/boards/scum/radio.c]: please add a comment that this is an arbitrary value for now

File: bsp/boards/scum/radio.c [hunk 24]
```diff
@@ -0,0 +1,312 @@
...
+
+    // clear variables
+    memset(&radio_vars,0,sizeof(radio_vars_t));
+    
+    // change state
+    radio_vars.state                = RADIOSTATE_STOPPED;
+    
+    // enable sfd done and send done interruptions of tranmission
+    // enable sfd done and receiving done interruptions of reception
+    RFCONTROLLER_REG__INT_CONFIG    = TX_SFD_DONE_INT_EN    |   \
```
Comment (twatteyne) [inline bsp/boards/scum/radio.c]: very cool! :+1:

File: bsp/boards/scum/radiotimer.c [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+#include "string.h"
+#include "radiotimer.h"
+#include "board.h"
+#include "bsp_timer.h"
+
+//=========================== define =======================================
+
+// ==== radio timer control bit
+
+#define RADIOTIMER_ENABLE                       0x01
```
Comment (twatteyne) [inline bsp/boards/scum/radiotimer.c]: Please start the name of this `#define` with the name of the register.

File: bsp/boards/scum/radiotimer.c [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+void radiotimer_setOverflowCb(radiotimer_compare_cbt cb) {
+    radiotimer_vars.overflow_cb    = cb;
+}
+
+void radiotimer_setCompareCb(radiotimer_compare_cbt cb) {
+    radiotimer_vars.compare_cb     = cb;
+}
+
+void radiotimer_setStartFrameCb(radiotimer_capture_cbt cb) {
+    while(1);
```
Comment (twatteyne) [inline bsp/boards/scum/radiotimer.c]: why `while(1)`? We took the VERY bad habbit to putting a `while(1)` to mean "I should never be here". But without JTAG, it just causes the board to halt without being able to debug. I would replace by a comment that says "should never be called".

File: bsp/boards/scum/uart.c (hunk 19 above)
Comment (twatteyne) [inline bsp/boards/scum/uart.c]: OK, thanks!

Comment (twatteyne) [inline bsp/boards/scum/uart.c]: could you add a comment to that effect in the enable/disable interrupt functions?

File: bsp/boards/scum/memory_map.h [hunk 27]
```diff
@@ -1,8 +1,8 @@
 /**
 \brief SCuM board's addresses of the memory-mapped registers .
 
-\author Sahar Misha     <smesri@berkeley.edu>,      August 2016.
-\author Tengfei Chang   <tengfei.chang@inria.fr>    August 2016.
+\author Sahar Misha Mesri   <smesri@berkeley.edu>,      August 2016.
```
Comment (twatteyne) [inline bsp/boards/scum/memory_map.h]: Where does Misha come from? Please just Sahar Mesri (https://www.linkedin.com/in/sahar-mesri-a8766435)

File: bsp/boards/scum/memory_map.h (hunk 12 above)
Comment (changtengfei) [inline bsp/boards/scum/memory_map.h]: They are moved to memory_map.h now.

File: bsp/boards/scum/uart.c (hunk 19 above)
Comment (changtengfei) [inline bsp/boards/scum/uart.c]: added.

File: bsp/boards/scum/board_info.h (hunk 21 above)
Comment (changtengfei) [inline bsp/boards/scum/board_info.h]: I added more comments on each operation.

File: bsp/boards/scum/radio.c (hunk 23 above)
Comment (changtengfei) [inline bsp/boards/scum/radio.c]: added

File: bsp/boards/scum/radio.c (hunk 24 above)
Comment (changtengfei) [inline bsp/boards/scum/radio.c]: :-)

File: bsp/boards/scum/board_info.h [hunk 28]
```diff
@@ -32,21 +32,19 @@ to return the board's description.
 
 // ==== SCuM RF timer specific 
 
-// since SCuM uses 500KHz, but the statck protocol is designed for 32KHz.
-// the following two marco are used to convert the couter value bewteen
-// two different frequency clocks. 500000/32768 is approximately 61/4.
-#define TIMER_COUTER_CONVERT_32K_TO_500K(value)    value*61/4
+/** 
+* since SCuM uses 500KHz, but the statck protocol is designed for 32KHz.
```
Comment (twatteyne) [inline bsp/boards/scum/board_info.h]: please fix typos in here

File: bsp/boards/scum/radiotimer.c (hunk 25 above)
Comment (changtengfei) [inline bsp/boards/scum/radiotimer.c]: changed and moved to memory_map.h file.

File: bsp/boards/scum/radiotimer.c (hunk 26 above)
Comment (changtengfei) [inline bsp/boards/scum/radiotimer.c]: replaced.

File: bsp/boards/scum/board_info.h [hunk 29]
```diff
@@ -33,8 +33,8 @@ to return the board's description.
 // ==== SCuM RF timer specific 
 
 /** 
-* since SCuM uses 500KHz, but the statck protocol is designed for 32KHz.
-* the following two marco are used to convert the couter value bewteen
+* since SCuM uses 500KHz, but the stack protocol is designed for 32KHz.
```
Comment (twatteyne) [inline bsp/boards/scum/board_info.h]: sorry to be a pain. KHz are "Kelvin Hertz". You probably mean kHz or "kilo-Hertz"

File: bsp/boards/scum/memory_map.h (hunk 27 above)
Comment (changtengfei) [inline bsp/boards/scum/memory_map.h]: Hmm, this is what shown in my gmail contact information.

File: bsp/boards/scum/board_info.h (hunk 29 above)
Comment (changtengfei) [inline bsp/boards/scum/board_info.h]: Glad to know :-)

File: bsp/boards/scum/leds.c [hunk 30]
```diff
@@ -64,26 +78,38 @@ uint8_t leds_sync_isOn() {
 
 // 3 <N14>
 void    leds_debug_on() {
+#ifndef DEBUGPINS_IN_USE
```
Comment (twatteyne) [inline bsp/boards/scum/leds.c]: I don't understand when I should define `DEBUGPINS_IN_USE`

Comment (changtengfei) [inline bsp/boards/scum/leds.c]: The leds.c file is not supposed to be committed. I will change it back when I finish the final port.

Comment (twatteyne) [inline bsp/boards/scum/leds.c]: ok

**Comment by twatteyne** [comment]
yayy!

**Pull request merged**